testing/endpoint.go

```diff
@@ -492,6 +493,85 @@ func (endpoint *Endpoint) TimeoutOnClose(packet channeltypes.Packet) error {
 	return endpoint.Chain.sendMsgs(timeoutOnCloseMsg)
 }
 
+// UpgradeChain performs a IBC client upgrade using the provided client state.
+// The chainID within the client state will have its revision number incremented by 1.
+// The counterparty client will be upgraded if it exists.
+func (endpoint *Endpoint) UpgradeChain(clientState *ibctmtypes.ClientState) error {
+	// increment revision number in chainID
+	revisionNumber := clienttypes.ParseChainID(clientState.ChainId)
+	newChainID, err := clienttypes.SetRevisionNumber(clientState.ChainId, revisionNumber+1)
+	if err != nil {
+		// current chainID is not in revision format
+		newChainID = clientState.ChainId + "-1"
+	}
+
+	clientState.ChainId = newChainID
+
+	// set upgraded client state
+	bz := clienttypes.MustMarshalClientState(endpoint.Chain.App.AppCodec(), clientState)
+	endpoint.Chain.GetSimApp().UpgradeKeeper.SetUpgradedClient(endpoint.Chain.GetContext(), endpoint.Chain.GetContext().BlockHeight(), bz)
+
+	// set upgraded consensus state
+	consensusState := &ibctmtypes.ConsensusState{
```

Comment:
Does this to include a root/app hash?

Reply:
Nope, the consensus state is set for the next block to be produced under the new chain id so the root hash cannot be known before hand